### ISSUE TITLE
Specified PGP signing subkey is not being used

### ISSUE THREAD

**Issue by atreyasha** [neomutt/neomutt#2980]
## Background

I like to keep my PGP primary key compartmentalized with different subkeys. Specifically, I use two signing subkeys; one for emails (subkey A) and one for git (subkey B), where subkey B is newer than subkey A.

##  Expected behavior

To specify the signing subkey that NeoMutt should use, I followed the instructions from NeoMutt's guide and added the following in my `muttrc`:

```
set pgp_sign_as = <subkey_A_id>
```

I would expect my emails to be signed with subkey A.

## Actual Behaviour

When analyzing the signatures of my sent emails, I observe they are all signed with subkey B (the newer one reserved for git).

## Attempted workarounds

I tried the following workarounds:

1. Appending an exclamation mark after the subkey ID, as recommended for git [here](https://stackoverflow.com/questions/46330629/signing-commits-in-git-uses-wrong-subkey)
2. I thought the [`pgp_ignore_subkeys`](https://neomutt.org/guide/reference#3-259- pgp_ignore_subkeys) option might be relevant and tried to add the following to my `muttrc`:

    ```
    set pgp_ignore_subkeys = no
    ```

None of these resulted in subkey A actually being used. It seems that NeoMutt ignores the specification for subkey A, infers the primary key and finds the latest signing subkey available for use; which would be subkey B in this case.

## Steps to Reproduce

1. Create a GPG key-pair with at least two signing subkeys
2. Specify the first (or older) signing subkey for signing in NeoMutt
3. Send a signed email to yourself
4. Download the automatically produced `signature.asc` file
5. Analyze the packets in the signature: `gpg --list-packets < signature.asc`
6. Compare the `issuer key ID` with your desired signing subkey; they should not be the same

## How often does this happen?

Always

## When did it start to happen?

I only tried this new PGP key configuration recently, so I am first encountering this issue now.

## NeoMutt Version

```
NeoMutt 20210205
Copyright (C) 1996-2020 Michael R. Elkins and others.
NeoMutt comes with ABSOLUTELY NO WARRANTY; for details type 'neomutt -vv'.
NeoMutt is free software, and you are welcome to redistribute it
under certain conditions; type 'neomutt -vv' for details.

System: Linux 5.10.44-1-lts (x86_64)
ncurses: ncurses 6.2.20200212 (compiled with 6.2.20200212)
libidn2: 2.3.1 (compiled with 2.3.0)
GPGME: 1.15.1
GnuTLS: 3.7.0
libnotmuch: 5.3.0
storage: kyotocabinet, gdbm, bdb, lmdb
compression: lz4, zlib, zstd

Configure options: --prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --gpgme --sqlite --autocrypt --lua --notmuch --gss --gnutls --sasl --with-ui=ncurses --with-idn2=/usr --disable-idn --idn2 --bdb --lmdb --kyotocabinet --gdbm --lz4 --zlib --zstd

Compilation CFLAGS: -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include -I/usr/include/lua5.3 -DNCURSES_WIDECHAR -I/include -I/usr/include/

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  +autocrypt +bkgdset +color +curs_set +fcntl -flock -fmemopen +futimens 
  +getaddrinfo +gnutls +gpgme +gss +hcache -homespool +idn +inotify 
  -locales_hack +lua +meta -mixmaster +nls +notmuch -openssl +pgp +regex +sasl 
  +smime +sqlite +start_color +sun_attachment +typeahead 
MAILPATH="/var/mail"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"

To learn more about NeoMutt, visit: https://neomutt.org
If you find a bug in NeoMutt, please raise an issue at:
    https://github.com/neomutt/neomutt/issues
or send an email to: <neomutt-devel@neomutt.org>
```

## Extra Info

* Operating System and its version: `Arch Linux 5.10.44-1-lts #1 SMP Wed, 16 Jun 2021 11:48:26 +0000 x86_64 GNU/Linux`
* Were you using multiple copies of NeoMutt at once: `No`
* Were you using 'screen' or 'tmux': `No`
* Is your email local (maildir) or remote (IMAP): `Local`

**Comment by atreyasha** [comment]
I found this old [post](https://chewt0y.wordpress.com/2004/09/27/using_openpgp_subkeys_in_mutt/) which seems to address the issue by setting the `pgp_sign_command` variable.

I have not tried this as yet, but I think one should be able to set the signing subkey via the `pgp_sign_as` variable without necessarily meddling with other variables.

**Comment by gahr** [comment]
> I found this old [post](https://chewt0y.wordpress.com/2004/09/27/using_openpgp_subkeys_in_mutt/) which seems to address the issue by setting the `pgp_sign_command` variable.

That'd work, but `pgp_sign_command` is used with the manual GPG config, where you'd set the `pgp_*_command` for everything, which we tend to discourage. We tend to encourage `crypt_use_gpgme`, where we use the GPGMe library instead of spawning gpg processes by hand. This is what I think you're using already.
 
> I have not tried this as yet, but I think one should be able to set the signing subkey via the `pgp_sign_as` variable without necessarily meddling with other variables.

I quickly double checked your steps and I can confirm that I  cannot pick a second subkey. I might be able to have a look, but I can't guarantee when :/

**Comment by gahr** [comment]
I have looked at the APIs and I can't seem to find a way to set a subkey as the signing key for a context. I think I'm not getting something.. I'll ask around.

**Comment by atreyasha** [comment]
> I quickly double checked your steps and I can confirm that I cannot pick the first (older) subkey

Good to know this can be reproduced.

> I can't seem to find a way to set a subkey as the signing key for a context

Like in a general context outside of `neomutt`? For me appending the subkey's ID with an exclamation mark works, for example: `gpg --detach-sign --output signature.asc --armor --local-user <subkey-ID>! <file>`

**Comment by gahr** [comment]
No, I mean a GPGME context, in the C API we use.

**Comment by gahr** [comment]
I have confirmed with gnupg developers that this cannot currently be done using GPGME APIs. I guess you could just split your keys of use the legacy `pgp_*_command` interface and set `crypt_use_gpgme=no`.

**Comment by gahr** [comment]
Upstream feature request: https://dev.gnupg.org/T5503

**Comment by atreyasha** [comment]
Awesome, thanks so much for following up upstream :)  Should we leave this issue open?

**Comment by gahr** [comment]
Yup, I'll keep this as an enhancement and implement our part once GPGME has its API released.
